aws-datastore/src/main/java/com/amplifyframework/datastore/syncengine/MutationLinkedList.java

```diff
@@ -0,0 +1,112 @@
+package com.amplifyframework.datastore.syncengine;
+
+import com.amplifyframework.core.model.Model;
+import java.util.HashMap;
+
+/*
+ * The {@link MutationLinkedList} is a LinkedHashMap like container , the goal of using this container is to
+ * achieve O(1) time complexity for both getting a {@link PendingMutation} and update an existing mutation with
+ * valid id.
+ */
+public class MutationLinkedList {
+
+    private HashMap<TimeBasedUuid, Node> mutationMap = new HashMap<>();
+    private Node dummyHead = new Node();
+    private Node dummyTail = new Node();
+
+    public MutationLinkedList() {
+        dummyHead.next = dummyTail;
+        dummyTail.prev = dummyHead;
+    }
+
+    public synchronized PendingMutation<? extends Model> nextMutationForModelId(String modelId) {
+        Node head = dummyHead.next;
+        if (head == dummyTail) {
+            return null;
+        }
+        while (head != dummyTail) {
+            if (head.mutation.getMutatedItem().getId().equals(modelId)) {
+                return head.mutation;
+            }
+            head = head.next;
+        }
+        return null;
+    }
+
+    public synchronized void remove(TimeBasedUuid id) {
+        if (mutationMap.containsKey(id)) {
+            Node removingNode = mutationMap.get(id);
+            Node removingPrev = removingNode.prev;
+            Node removingNext = removingNode.next;
+            //remove from map
+            mutationMap.remove(id);
+            //remove from current linkedlist
+            removingPrev.next = removingNext;
+            removingNext.prev = removingPrev;
+        }
+    }
+
+    public synchronized void clear() {
+        dummyHead.next = dummyTail;
+        dummyTail.prev = dummyHead;
+        mutationMap.clear();
+    }
+
+    public synchronized boolean isEmpty() {
+        return mutationMap.isEmpty();
+    }
+
+    public synchronized void addLast(PendingMutation<? extends Model> pendingMutation) {
+        //constructing a new node
+        Node pendingNode = new Node();
+        pendingNode.id = pendingMutation.getMutationId();
+        pendingNode.mutation = pendingMutation;
+        //insert the new node into the current linkedlist
+        Node addingPrev = dummyTail.prev;
+        addingPrev.next = pendingNode;
+        dummyTail.prev = pendingNode;
+        pendingNode.prev = addingPrev;
+        pendingNode.next = dummyTail;
+        //put the mutation into the mutationmap
+        mutationMap.put(pendingMutation.getMutationId(), pendingNode);
```

Comment:
Instead of maintaining your own head and tail, can you use a linked list/map or some kind directly? "Mutation linked list _has a_ (some type of) list, composed into it?"

Reply:
@jamesonwilliams i was thinking about using LinkedListHashMap which internally does the same thing as my new data structure , but LinkedListHashMap doesnt expose the api to iterate the linkedlist from head to tail. Also same for LinkedList java standard implementation , if you try to remove it it will traverse through the entire LinkedList and find the target which is again O(N) time complexity.
This data structure is actually inspired by a leetcode question i did before:
https://leetcode.com/problems/lru-cache/

Its for LRU cache but does the same thing.